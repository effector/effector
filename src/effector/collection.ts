export function forIn<T, Key extends string = string>(
  obj: Record<Key, T>,
  cb: (value: T, key: Key) => void,
) {
  for (const key in obj) {
    cb(obj[key], key)
  }
}

export const includes = <T>(list: T[], item: T) => list.includes(item)

export const removeItem = <T>(list: T[], item: T) => {
  const pos = list.indexOf(item)
  if (pos !== -1) {
    list.splice(pos, 1)
  }
}

export const add = <T>(list: T[], item: T) => list.push(item)

export function forEach<T>(
  list: T[],
  fn: (item: T, index: number, list: T[]) => void,
): void
export function forEach<K, T>(
  list: Map<K, T>,
  fn: (item: T, key: K) => void,
): void
export function forEach<T>(list: Set<T>, fn: (item: T) => void): void
export function forEach(list: any, fn: Function) {
  list.forEach(fn)
}

/** generalized visitor with duplicates check */
export function traverse<T>(
  startItem: T,
  cb: (item: T, visit: (item: T) => void) => void,
) {
  const visited = new Set<T>()
  const stack = [startItem]
  const visit = (item: T) => stack.push(item)
  let item: T | void
  while ((item = stack.shift())) {
    if (!visited.has(item)) {
      visited.add(item)
      cb(item, visit)
    }
  }
}
