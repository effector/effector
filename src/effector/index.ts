/*
 * People assume that time is a strict progression of cause to effect,
 * but actually from a non-linear, non-subjective viewpoint -
 * it's more like a big ball
 */

export {attach} from './attach'
export {clearNode} from './clearNode'
export {combine} from './combine'
export {createApi} from './createApi'
export {createDomain} from './createDomain'
export {createEffect} from './createEffect'
export {createNode} from './createNode'
export {createStore, createEvent} from './createUnit'
export {forward} from './forward'
export {fromObservable} from './fromObservable'
export {guard} from './guard'
export {is} from './is'
export {launch, setInspector} from './kernel'
export {merge} from './merge'
export {restore} from './restore'
export {sample} from './sample'
export {setStoreName} from './naming'
export {split} from './split'
export {step} from './step'
export {version} from './flags'
export {withRegion, withFactory, setGraphInspector} from './region'
export {hydrate, serialize, scopeBind, fork, allSettled} from './fork'
export {createWatch} from './createWatch'