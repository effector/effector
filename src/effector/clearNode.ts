import {Node, NodeUnit} from './index.h'
import {getGraph, getOwners, getLinks, getSubscribers} from './getter'
import {is} from './is'
import {removeItem} from './collection'
import {CROSSLINK} from './tag'

const removeFromNode = (currentNode: Node, targetNode: Node) => {
  removeItem(currentNode.next, targetNode)
  removeItem(getOwners(currentNode), targetNode)
  removeItem(getLinks(currentNode), targetNode)
}
const clearNodeNormalized = (
  targetNode: Node,
  deep: boolean,
  isDomainUnit: boolean,
) => {
  targetNode.next.length = 0
  targetNode.seq.length = 0
  //@ts-expect-error
  targetNode.scope = null
  let currentNode
  let list = getLinks(targetNode)
  while ((currentNode = list.pop())) {
    removeFromNode(currentNode, targetNode)
    if (
      deep ||
      (isDomainUnit && targetNode.meta.op !== 'sample') ||
      currentNode.family.type === CROSSLINK
    ) {
      clearNodeNormalized(
        currentNode,
        deep,
        currentNode.meta.op !== 'on' && isDomainUnit,
      )
    }
  }
  list = getOwners(targetNode)
  while ((currentNode = list.pop())) {
    removeFromNode(currentNode, targetNode)
    if (isDomainUnit && currentNode.family.type === CROSSLINK) {
      clearNodeNormalized(
        currentNode,
        deep,
        currentNode.meta.op !== 'on' && isDomainUnit,
      )
    }
  }
}
const clearMap = (map: Map<any, any> | Set<any>) => map.clear()
export const clearNode = (
  graphite: NodeUnit,
  {
    deep,
  }: {
    deep?: boolean
  } = {},
) => {
  let isDomainUnit = false
  //@ts-expect-error
  if (graphite.ownerSet) graphite.ownerSet.delete(graphite)
  if (is.store(graphite)) {
    clearMap(getSubscribers(graphite))
  } else if (is.domain(graphite)) {
    isDomainUnit = true
    const history = graphite.history
    clearMap(history.events)
    clearMap(history.effects)
    clearMap(history.stores)
    clearMap(history.domains)
  }
  clearNodeNormalized(getGraph(graphite), !!deep, isDomainUnit)
}
