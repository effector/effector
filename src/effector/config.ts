import {forIn} from './collection'
import {assertObject, isObject, isVoid} from './is'

export function processArgsToConfig(
  arg: any,
  singleArgument: true,
): [any, any | void]
export function processArgsToConfig(args: any[]): [any[], any | void]
export function processArgsToConfig(
  args: any[],
  singleArgument?: boolean,
): [any[], any | void] {
  const rawConfig = singleArgument ? args : args[0]
  assertObject(rawConfig)
  let metadata = rawConfig.or
  const childConfig = rawConfig.and
  if (childConfig) {
    const unwrappedNestedValue = singleArgument ? childConfig : childConfig[0]
    /**
     * if there is no "and" field then we reached the leaf of the tree
     * and this is an original user-defined argument
     *
     * note that in this case we're returning all arguments, not the only one been unwrapped
     **/
    if (!isObject(unwrappedNestedValue) || !('and' in unwrappedNestedValue)) {
      args = childConfig
    } else {
      //@ts-expect-error
      const nested = processArgsToConfig(childConfig, singleArgument)

      args = nested[0]
      metadata = {...metadata, ...nested[1]}
    }
  }
  return [args, metadata]
}

/**
processed fields:

'name',
'sid',
'loc',
'handler',
'updateFilter',
'parent',
'serialize',
'named',
'derived',
*/
export const flattenConfig = (part: any, config: Record<string, any> = {}) => {
  if (isObject(part)) {
    flattenConfig(part.or, config)
    forIn(part, (value, field) => {
      if (!isVoid(value) && field !== 'or' && field !== 'and') {
        config[field] = value
      }
    })
    flattenConfig(part.and, config)
  }
  return config
}
