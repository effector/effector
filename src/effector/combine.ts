import type {Store} from './unit.h'
import {createStore, requireExplicitSkipVoidMessage} from './createUnit'
import {createStateRef, addRefOp} from './stateRef'
import {mov, calc, read, userFnCall} from './step'
import {processArgsToConfig} from './config'
import {getStoreState, setMeta} from './getter'
import {is, isFunction, isObject, isVoid} from './is'
import {generateErrorTitle, unitObjectName} from './naming'
import {createLinkNode} from './forward'
import {assert, deprecate} from './throw'
import {readTemplate} from './region'
import {forIn} from './collection'
import {MAP, REG_A, VALUE} from './tag'
import {applyTemplate} from './template'
import type {Config} from './index.h'

export function combine(...args: any[]): Store<any> {
  let handler
  let stores
  let config
  ;[args, config] = processArgsToConfig(args)
  const errorTitle = generateErrorTitle('combine', config)
  // skipVoid support, to be removed in effector 24
  const maybeExtConfig = args[args.length - 1]
  const isExtendedConfig = !is.store(maybeExtConfig) && isObject(maybeExtConfig)
  const extConfig = isExtendedConfig && maybeExtConfig
  const rawHandler = isExtendedConfig ? args[args.length - 2] : maybeExtConfig
  if (isFunction(rawHandler)) {
    stores = args.slice(0, isExtendedConfig ? -2 : -1)
    handler = rawHandler
  } else {
    stores = args
  }

  let structStoreShape
  let shapeReady
  if (stores.length === 1) {
    const obj = stores[0]
    /*
      without edge case combine(Color, (Color) => '~')
      */
    if (!is.store(obj)) {
      /*
      case combine([R,G,B], ([R,G,B]) => '~')
      case combine({R,G,B}, ({R,G,B}) => '~')

      edge case combine([Color], ([Color]) => '~')
      edge case combine({Color}, ({Color}) => '~')

      edge case combine([R,G,B])
      edge case combine({R,G,B})

      edge case combine([Color])
      edge case combine({Color})
      */
      structStoreShape = obj
      shapeReady = true
    }
  }
  let noArraySpread: boolean | void
  if (!shapeReady) {
    /*
    case combine(R,G,B, (R,G,B) => '~')
    */
    structStoreShape = stores
    /*
    without edge case combine(R,G,B)
    without edge case combine(Color)
    */
    if (handler) {
      noArraySpread = true
      const fn = handler
      handler = (list: any[]) => fn(...list)
    }
  }
  assert(isObject(structStoreShape), `${errorTitle}: shape should be an object`)
  return storeCombination(
    Array.isArray(structStoreShape),
    !noArraySpread,
    structStoreShape,
    config,
    handler,
    extConfig,
  )
}

const storeCombination = (
  isArray: boolean,
  needSpread: boolean,
  obj: any,
  config?: Config,
  fn?: (upd: any) => any,
  extConfig?: {skipVoid?: boolean},
) => {
  const errorTitle = generateErrorTitle('combine', config)
  const clone = isArray ? (list: any) => [...list] : (obj: any) => ({...obj})
  const defaultState: Record<string, any> = isArray ? [] : {}

  const stateNew = clone(defaultState)
  const rawShape = createStateRef(stateNew)
  const isFresh = createStateRef(true)
  rawShape.type = isArray ? 'list' : 'shape'
  rawShape.noInit = true
  applyTemplate('combineBase', rawShape, isFresh)
  const store = createStore(stateNew, {
    name: unitObjectName(obj),
    derived: true,
    ...extConfig,
    and: config,
  })
  const storeStateRef = getStoreState(store)
  storeStateRef.noInit = true
  setMeta(store, 'isCombine', true)
  const rawShapeReader = read(rawShape)
  /**
   * usual ref reading has very high priority, which leads to data races
   * ref reading for combine should have same "barrier" priority but without batching
   * (thats why order has no "barrierID" field, which assume batching)
   **/
  rawShapeReader.order = {priority: 'barrier'}
  /**
   * Soft store reading is required for
   * setting target store as inited in scope
   * for preventing retriggering issues
   **/
  const softReader = mov({
    store: storeStateRef,
    to: 'b',
    priority: 'read',
  })
  softReader.data.softRead = true
  const node = [
    calc((upd, _, stack) => {
      if (stack.scope && !stack.scope.reg[rawShape.id]) {
        stack.c = true
      }
      return upd
    }),
    rawShapeReader,
    mov({store: isFresh, to: 'b'}),
    calc((upd, {key}, reg) => {
      if (reg.c || upd !== reg.a[key]) {
        if (needSpread && reg.b) {
          reg.a = clone(reg.a)
        }
        reg.a[key] = upd
        return true
      }
    }, true),
    mov({from: REG_A, target: rawShape}),
    mov({from: VALUE, store: false, target: isFresh}),
    mov({
      from: VALUE,
      store: true,
      target: isFresh,
      priority: 'barrier',
      batch: true,
    }),
    /**
     * `read` with `sampler` priority is used to prevent cases,
     *  where `combine` triggers are duplicated
     *
     *  basically, this makes `sample` and `combine` priorities equal
     */
    read(rawShape, true, true),
    fn && userFnCall(),
    softReader,
  ]
  forIn(obj, (child: Store<any> | any, key) => {
    if (!is.store(child)) {
      assert(
        !is.unit(child) && !isVoid(child),
        `combine expects a store in a field ${key}`,
        errorTitle,
      )
      stateNew[key] = defaultState[key] = child
      return
    }
    defaultState[key] = child.defaultState
    stateNew[key] = child.getState()
    const linkNode = createLinkNode(child, store, node, 'combine', fn)
    linkNode.scope.key = key
    const childRef = getStoreState(child)
    addRefOp(rawShape, {type: 'field', field: key, from: childRef})
    applyTemplate('combineField', childRef, linkNode)
  })

  store.defaultShape = obj
  addRefOp(storeStateRef, {
    type: MAP,
    from: rawShape,
    fn,
  })
  if (!readTemplate()) {
    if (fn) {
      const computedValue = fn(stateNew)

      if (isVoid(computedValue) && (!extConfig || !('skipVoid' in extConfig))) {
        console.error(`${errorTitle}: ${requireExplicitSkipVoidMessage}`)
      }

      storeStateRef.current = computedValue
      storeStateRef.initial = computedValue
      store.defaultState = computedValue
    } else {
      store.defaultState = defaultState
    }
  }
  return store
}
