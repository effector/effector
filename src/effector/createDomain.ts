import type {Domain} from './unit.h'
import {own} from './own'
import {createNode} from './createNode'
import type {Config, NodeUnit} from './index.h'
import {
  createEvent,
  createNamedEvent,
  createStore,
  initUnit,
} from './createUnit'
import {createEffect} from './createEffect'
import {createLinkNode} from './forward'
import {add, forEach, forIn} from './collection'
import {getGraph, getParent} from './getter'
import {DOMAIN} from './tag'
import {launch} from './kernel'
import {calc} from './step'
import {flattenConfig} from './config'

export function createDomain(nameOrConfig: any, maybeConfig?: any): Domain {
  const config = flattenConfig({
    or: maybeConfig,
    and: typeof nameOrConfig === 'string' ? {name: nameOrConfig} : nameOrConfig,
  }) as any

  const node = createNode({
    family: {type: DOMAIN},
    regional: true,
    parent: config?.domain || config?.parent,
  })

  const domain = {
    history: {},
    graphite: node,
    hooks: {},
  } as Domain

  node.meta = initUnit(DOMAIN, domain, {
    parent: config?.domain || config?.parent,
    or: {...config, derived: true},
  })

  forIn(
    {
      Event: createEvent,
      Effect: createEffect,
      Store: createStore,
      Domain: createDomain,
    },
    (factory, tag) => {
      const lowerCaseTag = tag.toLowerCase() as
        | 'event'
        | 'effect'
        | 'store'
        | 'domain'

      const onCreateUnit = createNamedEvent(`on${tag}`)
      domain.hooks[lowerCaseTag] = onCreateUnit

      const history = new Set<any>()
      domain.history[`${lowerCaseTag}s`] = history

      onCreateUnit.create = unit => {
        launch(onCreateUnit, unit)
        return unit
      }
      add(
        getGraph(onCreateUnit).seq,
        calc((upd, _, stack) => {
          stack.scope = null
          return upd
        }),
      )
      onCreateUnit.watch(unit => {
        own(domain, [unit])
        history.add(unit)
        if (!unit.ownerSet) unit.ownerSet = history
        if (!getParent(unit)) unit.parent = domain
      })
      own(domain, [onCreateUnit])

      domain[`onCreate${tag}`] = (hook: (unit: any) => any) => {
        forEach(history, hook)
        return onCreateUnit.watch(hook)
      }
      domain[`create${tag}`] = domain[lowerCaseTag] = (
        nameOrConfig: any,
        rawConfig?: Config,
      ) => {
        const config = flattenConfig({and: rawConfig, or: nameOrConfig})
        if (config?.domain) {
          // @ts-expect-error complicated factory type
          return factory(nameOrConfig, rawConfig)
        }
        // @ts-expect-error complicated factory type
        return onCreateUnit(factory(nameOrConfig, {parent: domain, or: config}))
      }
    },
  )

  const parent = getParent(domain)
  if (parent) {
    forIn(domain.hooks, (from: NodeUnit, key) =>
      createLinkNode(from, parent.hooks[key]),
    )
  }
  if (config?.domain) {
    config.domain.hooks.domain(domain)
  }
  return domain
}
