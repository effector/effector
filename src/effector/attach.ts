import type {Domain} from './unit.h'
import {combine} from './combine'
import {createEffect, onSettled, runFn} from './createEffect'
import {applyParentHook} from './createUnit'
import {processArgsToConfig} from './config'
import {
  getGraph,
  getParent,
  getStoreState,
  setMeta,
  getCompositeName,
} from './getter'
import {own} from './own'
import {is, isVoid} from './is'
import {read, calc} from './step'
import {launch} from './kernel'
import {EFFECT} from './tag'
import {createName, generateErrorTitle} from './naming'
import {assert} from './throw'
import {Cmd, Node, Stack} from './index.h'

export function attach(config: any) {
  let injected
  ;[config, injected] = processArgsToConfig(config, true)
  const errorTitle = generateErrorTitle('attach', injected)
  let {source, effect, mapParams, domain} = config
  if (is.effect(effect)) {
    assert(
      isVoid(domain),
      '`domain` can only be used with a plain function',
      errorTitle,
    )
  }
  const attached = createEffect(config, injected)
  setMeta(attached, 'attached', true)
  const {runner} = getGraph(attached).scope as {runner: Node}
  let runnerSteps: Array<Cmd>
  const runnerFnStep = (upd: any, _: any, stack: Stack) => {
    const {params, req, handler} = upd
    const anyway = attached.finally
    const rj = onSettled(params, req, false, anyway, stack)
    const sourceData = stack.a
    const isEffectHandler = is.effect(handler)
    let ok = true
    let computedParams
    if (mapParams) {
      ;[ok, computedParams] = runFn(mapParams, rj, [params, sourceData])
    } else {
      computedParams = source && isEffectHandler ? sourceData : params
    }
    if (ok) {
      if (isEffectHandler) {
        launch({
          target: handler as any,
          params: {
            params: computedParams,
            req: {
              rs: onSettled(params, req, true, anyway, stack),
              rj,
            },
          },
          page: stack.page,
          defer: true,
          meta: stack.meta,
        })
      } else {
        upd.args = [sourceData, computedParams]
        return true
      }
    }
  }
  if (source) {
    runner.scope.runnerFn = runnerFnStep
    let state
    if (is.store(source)) {
      state = source
      own(state, [attached])
    } else {
      state = combine(source)
      own(attached, [state])
    }
    runnerSteps = [
      /**
       * effect computation should run in effect queue,
       * but reading state too early leads to data races
       * so we add additional empty step
       **/
      calc(upd => upd, false, true),
      read(getStoreState(state)),
    ]
    /**
     * in simple effect, seq consists of two steps
     * first prepares handler and the second actually calls it
     *
     * we delete priority, as we already in effect queue
     * so additional delay is not needed
     *
     * curiously, presence of effect here leads to stale values
     * because if effect step is not first in a sequence
     * it will be placed in queue
     * to prevent data races: effects should be called in order,
     * but before we run effect in a queue, we should run
     * more important tasks at first, so it should execute like this:
     *
     *    pure queue: [node [mov foo reg_a, effect a]]
     *    effect queue: [node [effect b]]
     *
     *    order: mov foo reg_a, effect b, effect a
     *
     * but when node is executed, everything in it will run sequentially
     * so kernel will find "effect a" earlier than already queued "effect b"
     *
     * this behavior leads to state reading happened before another effect handlers
     * execution and in this case we dont want that, as it leads
     * to reading values which are going to change
     */
    delete runner.seq[1].order
  } else {
    runnerSteps = [calc(runnerFnStep, true, true)]
  }
  runner.seq.splice(1, 0, ...runnerSteps)
  attached.use(effect)
  const parentDomain: Domain | void = getParent(effect)
  if (parentDomain) {
    Object.assign(
      getCompositeName(attached),
      createName(attached.shortName, parentDomain),
    )
    attached.defaultConfig.parent = parentDomain
  }
  applyParentHook(effect, attached, EFFECT)
  return attached
}
