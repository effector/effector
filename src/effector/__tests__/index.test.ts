import {from} from 'most'

import {
  createStore,
  Store,
  combine,
  createEvent,
  createEffect,
  Event,
  EventCallable,
  forward,
  split,
  guard,
  sample,
} from 'effector'
//@ts-expect-error
import {show} from 'effector/fixtures/showstep'
import {argumentHistory} from 'effector/fixtures'

test('graphite', () => {
  const fn = jest.fn()
  const fn1 = jest.fn()
  const foo: EventCallable<number> = createEvent('foo')
  const bar = foo.map(x => (fn(x), x + 1))
  const store1: Store<string> = createStore('foo').on(bar, (state, bar) =>
    [state, bar].join(' | '),
  )
  const store2 = store1.map(e => e.length)
  const unsub1 = bar.watch(fn1)
  foo(0)
  foo(10)
  expect(fn).toHaveBeenCalledTimes(2)
  expect(store1.getState()).toBe('foo | 1 | 11')
  //@ts-expect-error
  const showBar = show(bar.graphite)

  expect(showBar).toMatchSnapshot('show bar')
  unsub1()
  //@ts-expect-error
  const showBar2 = show(bar.graphite)
  expect(showBar2).toMatchSnapshot('show bar')
  foo(100)
  expect(fn1).toHaveBeenCalledTimes(2)
  //@ts-expect-error
  const showFoo = show(foo.graphite)
  //@ts-expect-error
  const showStore = show(store1.graphite)
  //@ts-expect-error
  const showStore2 = show(store2.graphite)
  expect(showFoo).toMatchSnapshot('show foo')
  expect(showStore).toMatchSnapshot('show store1')
  expect(showStore2).toMatchSnapshot('show store2')
})

test('showcase', () => {
  const fn = jest.fn()
  const foo = createEvent('foo')
  const bar = createEvent('bar')

  const a = createStore(1)
  const b = createStore(2)
  const bigStore = combine({a, b})
  const mapped = bigStore.map(s => s.a)

  a.on(foo, n => n + 1)
  b.on(bar, n => n + 1)

  mapped.watch(fn)

  foo()
  foo()

  bar()
  bar()
  bar()
  bar()

  //@ts-expect-error
  expect(show(a.graphite)).toMatchSnapshot('store a')
  //@ts-expect-error
  expect(show(foo.graphite)).toMatchSnapshot('event foo')
  //@ts-expect-error
  expect(show(mapped.graphite)).toMatchSnapshot('mapped')
  expect(fn).toHaveBeenCalledTimes(3)
  const first = createStore('s')
  const second = createStore('h')
  const third = createStore('i')
  const status = combine(first, second, third, (a, b, c) => [a, b, c].join(''))
  expect(status.getState()).toBe('shi')
})

describe('symbol-observable support', () => {
  test('from(store)', async () => {
    const fn = jest.fn()
    expect(() => {
      from(createStore(0))
    }).not.toThrow()
    const store1 = createStore(-1)
    const ev1 = createEvent<string>()
    const ev2 = createEvent<string>()
    const store1$ = from(store1)
    store1$.observe(fn)
    store1.on(ev1, state => state + 1)
    ev1('foo')
    ev1('bar')
    ev1('baz')
    ev2('should ignore')

    expect(argumentHistory(fn)).toEqual([-1, 0, 1, 2])
    expect(fn).toHaveBeenCalledTimes(4)
    expect(store1.getState()).toBe(2)
  })
  describe('from(effect)', () => {
    test('without implementation', async () => {
      const fn = jest.fn()
      expect(() => {
        from(createEffect())
      }).not.toThrow()
      const ev1 = createEffect((_: number) => {})
      const ev2 = createEffect((_: string) => {})
      const ev1$ = from(ev1)
      ev1$.observe(fn)
      ev1(0)
      ev1(1)
      ev1(2)
      ev2('should ignore')
      expect(fn).toHaveBeenCalledTimes(3)

      expect(argumentHistory(fn)).toEqual([0, 1, 2])
    })

    test('with implementation', async () => {
      const fn = jest.fn()
      expect(() => {
        async function impl() {}
        const eff1 = createEffect()
        eff1.use(impl)
        from(eff1)
      }).not.toThrow()
      const ev1 = createEffect((_: number) => {})
      const ev2 = createEffect((_: string) => {})
      async function impl() {}
      ev1.use(impl)
      const ev1$ = from(ev1)
      ev1$.observe(fn)
      ev1(0)
      ev1(1)
      ev1(2)
      ev2('should ignore')
      expect(fn).toHaveBeenCalledTimes(3)

      expect(argumentHistory(fn)).toEqual([0, 1, 2])
    })
  })
})

test('attt', () => {
  const state = createStore({
    foo: 1,
    bar: [0],
  })
  const e1 = createEvent<string>()
  expect(state.getState()).toMatchObject({
    foo: 1,
    bar: [0],
  })
  state.on(e1, (state, payload) => ({
    foo: parseInt(payload, 36),
    bar: [...state.bar, parseInt(payload, 36)],
  }))
  e1('10')
  expect(state.getState()).toMatchObject({
    foo: 36,
    bar: [0, 36],
  })
  e1('11')
  expect(state.getState()).toMatchObject({
    foo: 37,
    bar: [0, 36, 37],
  })
})

test('createStore', () => {
  const counter = createStore(0)
  const text = createStore('')
  const store = combine({counter, text, foo: 'bar'})
  expect(store.getState()).toMatchObject({counter: 0, text: '', foo: 'bar'})
})

test('store.watch', () => {
  const click = createEvent<string | void>()
  const store1 = createStore(-1)
  const fn1 = jest.fn()
  const fn2 = jest.fn()
  store1.watch(fn1)
  store1.on(click, (state, e) => (fn2(state, e), state))
  click()
  click('a')
  click('b')
  expect(store1.getState()).toBe(-1)

  expect(fn1).not.toHaveBeenCalledTimes(2)
  expect(fn2).toHaveBeenCalledTimes(3)

  expect(fn1).toHaveBeenCalledTimes(1)
  expect(fn2.mock.calls).toEqual([
    [-1, undefined],
    [-1, 'a'],
    [-1, 'b'],
  ])
  expect(argumentHistory(fn1)).toEqual([-1])
})
