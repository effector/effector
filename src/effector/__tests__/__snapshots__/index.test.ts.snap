// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphite: show bar 1`] = `
"Node {
  seq: [],
  next: [
    Node {
      seq: [
        mov: store a,
        compute: callARegStack
      ],
      next: [
        Node {
          seq: [
            compute: Function,
            mov: store a,
            filter: Function,
            mov: stack store
          ],
          next: [
            Node {seq: [], next: []},
            Node {
              seq: [
                mov: store a,
                compute: callStack
              ],
              next: [
                Node {
                  seq: [
                    compute: Function,
                    mov: store a,
                    filter: Function,
                    mov: stack store
                  ],
                  next: [
                    Node {seq: [], next: []}
                  ]
                }
              ]
            }
          ]
        }
      ]
    },
    Node {
      seq: [
        run: effect
      ],
      next: []
    }
  ]
}"
`;

exports[`graphite: show bar 2`] = `
"Node {
  seq: [],
  next: [
    Node {
      seq: [
        mov: store a,
        compute: callARegStack
      ],
      next: [
        Node {
          seq: [
            compute: Function,
            mov: store a,
            filter: Function,
            mov: stack store
          ],
          next: [
            Node {seq: [], next: []},
            Node {
              seq: [
                mov: store a,
                compute: callStack
              ],
              next: [
                Node {
                  seq: [
                    compute: Function,
                    mov: store a,
                    filter: Function,
                    mov: stack store
                  ],
                  next: [
                    Node {seq: [], next: []}
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}"
`;

exports[`graphite: show foo 1`] = `
"Node {
  seq: [],
  next: [
    Node {
      seq: [
        compute: callStack
      ],
      next: [
        Node {
          seq: [],
          next: [
            Node {
              seq: [
                mov: store a,
                compute: callARegStack
              ],
              next: [
                Node {
                  seq: [
                    compute: Function,
                    mov: store a,
                    filter: Function,
                    mov: stack store
                  ],
                  next: [
                    Node {seq: [], next: []},
                    Node {
                      seq: [
                        mov: store a,
                        compute: callStack
                      ],
                      next: [
                        Node {
                          seq: [
                            compute: Function,
                            mov: store a,
                            filter: Function,
                            mov: stack store
                          ],
                          next: [
                            Node {seq: [], next: []}
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}"
`;

exports[`graphite: show store1 1`] = `
"Node {
  seq: [
    compute: Function,
    mov: store a,
    filter: Function,
    mov: stack store
  ],
  next: [
    Node {seq: [], next: []},
    Node {
      seq: [
        mov: store a,
        compute: callStack
      ],
      next: [
        Node {
          seq: [
            compute: Function,
            mov: store a,
            filter: Function,
            mov: stack store
          ],
          next: [
            Node {seq: [], next: []}
          ]
        }
      ]
    }
  ]
}"
`;

exports[`graphite: show store2 1`] = `
"Node {
  seq: [
    compute: Function,
    mov: store a,
    filter: Function,
    mov: stack store
  ],
  next: [
    Node {seq: [], next: []}
  ]
}"
`;

exports[`showcase: event foo 1`] = `
"Node {
  seq: [],
  next: [
    Node {
      seq: [
        mov: store a,
        compute: callARegStack
      ],
      next: [
        Node {
          seq: [
            compute: Function,
            mov: store a,
            filter: Function,
            mov: stack store
          ],
          next: [
            Node {seq: [], next: []},
            Node {
              seq: [
                compute: Function,
                mov: store a,
                mov: store b,
                filter: Function,
                mov: a store,
                mov: value store,
                mov: value store,
                mov: store stack,
                mov: store b
              ],
              next: [
                Node {
                  seq: [
                    compute: Function,
                    mov: store a,
                    filter: Function,
                    mov: stack store
                  ],
                  next: [
                    Node {seq: [], next: []},
                    Node {
                      seq: [
                        mov: store a,
                        compute: callStack
                      ],
                      next: [
                        Node {
                          seq: [
                            compute: Function,
                            mov: store a,
                            filter: Function,
                            mov: stack store
                          ],
                          next: [
                            Node {seq: [], next: []},
                            Node {
                              seq: [
                                run: effect
                              ],
                              next: []
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}"
`;

exports[`showcase: mapped 1`] = `
"Node {
  seq: [
    compute: Function,
    mov: store a,
    filter: Function,
    mov: stack store
  ],
  next: [
    Node {seq: [], next: []},
    Node {
      seq: [
        run: effect
      ],
      next: []
    }
  ]
}"
`;

exports[`showcase: store a 1`] = `
"Node {
  seq: [
    compute: Function,
    mov: store a,
    filter: Function,
    mov: stack store
  ],
  next: [
    Node {seq: [], next: []},
    Node {
      seq: [
        compute: Function,
        mov: store a,
        mov: store b,
        filter: Function,
        mov: a store,
        mov: value store,
        mov: value store,
        mov: store stack,
        mov: store b
      ],
      next: [
        Node {
          seq: [
            compute: Function,
            mov: store a,
            filter: Function,
            mov: stack store
          ],
          next: [
            Node {seq: [], next: []},
            Node {
              seq: [
                mov: store a,
                compute: callStack
              ],
              next: [
                Node {
                  seq: [
                    compute: Function,
                    mov: store a,
                    filter: Function,
                    mov: stack store
                  ],
                  next: [
                    Node {seq: [], next: []},
                    Node {
                      seq: [
                        run: effect
                      ],
                      next: []
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}"
`;
